packages/binding-mqtt/test/mqtt-client-subscribe-test.integration.ts

```diff
@@ -155,4 +155,432 @@ describe("MQTT client implementation", () => {
             });
         });
     }).timeout(20000);
+
+    it("should write property (string)", (done: Mocha.Done) => {
+        brokerServer = new MqttBrokerServer({ uri: brokerUri, selfHost: true });
+        servient.addServer(brokerServer);
+
+        servient.addClientFactory(new MqttClientFactory());
+
+        servient.start().then((WoT) => {
+            expect(brokerServer.getPort()).to.equal(brokerPort);
+            expect(brokerServer.getAddress()).to.equal(brokerAddress);
+
+            const propertyNumber = Math.floor(Math.random() * 1000000);
+            const propertyName: string = "property" + propertyNumber;
+            const properties: { [key: string]: Record<string, unknown> } = {};
+            properties[propertyName] = {
+                type: "string",
+                observable: true,
+                readOnly: false,
+                writeOnly: true,
+            };
+
+            WoT.produce({
+                title: "TestWoTMQTT",
+                properties: properties,
+            }).then((thing) => {
+                let receivedInput = "";
+
+                thing.setPropertyWriteHandler(propertyName, async (inputData) => {
+                    receivedInput = (await inputData.value()) as string;
+                });
+
+                thing.expose().then(() => {
+                    info(`Exposed ${thing.getThingDescription().title}`);
+
+                    WoT.consume(thing.getThingDescription()).then((client) => {
+                        const input = "writeProperty";
+
+                        client
+                            .writeProperty(propertyName, input)
+                            .then(() => {
+                                setTimeout(() => {
+                                    expect(input).to.equal(receivedInput);
+                                    thing.destroy().then(() => done());
+                                }, 100);
```

Comment:
I'm not a fan of timeouts and I try to use them as less as possible. Using a `qos` up to 2 should make the publishing synchronous and therefore you should be able to verify the execution of the writing operation without timeouts.

Reply:
I don't know if qos works in the implementation because I haven't seen any dependent logic to qos, could you explain it more how it works? Also If the mqtt broker server is slow would it matter we have qos 2 or not? We could get confirmation that we published our message but we might not know how long the process will take in the broker server.

Reply:
@relu91 actually even with qos 2 and the function being synchronous, the Thing can take a while to process the input it received. I think that this timeout is necessary but can be smaller. However, running the action takes too long and 20ms was not enough at [this commit.](https://github.com/eclipse-thingweb/node-wot/pull/992/commits/ae91065e408cb2f0e4f430912dedc74fb11c611f#diff-c046a35f74347f972fe11252680105f35586ad7ef2cfbc6d2fd6376e105708c8R201).

Regarding making it qos 2, I think that this is a missing feature. At this [line](https://github.com/eclipse-thingweb/node-wot/blob/badc223b330d588f336435589270ffb0a41ad1a7/packages/binding-mqtt/src/mqtt-client.ts#L102), we are not using qos of the forms at all.

Ideally, this test should happen inside the write handler. Since it is a callback, we are not sure if it will work but @hasanheroglu will give it a try.

Reply:
Using QOS 2 makes it equivalent to an HTTP POST/PUT, it is true that the thing could take a while but that's why there is a global timeout defined for each test. It should be something exceptional that would not happen in the circumstances of this test (cause usually, everything is local). 

About QoS 2 I was thinking that we had that, but the idea of testing it inside the writeHandler is not completely bad (even though is not a proper e2e test, meaning that we could mess up the writing of the payload somehow).

Reply:
`expect` and `done` seems to work in a callback. New tests pass.